### PR TITLE
add preferredDuringSchedulingIgnoredDuringExecution option

### PR DESCRIPTION
**Pull request by krudskoy**

### CODE REVIEW

**Comment by aveshagarwal** [comment]
Can one of the admins verify this patch?

**Comment by krudskoy** [comment]
Hello @ravisantoshgudimetla, I can't access the link [104.196.33.47:8080/job/descheduler-CI/165/](url) and see what happend with default test. Does the service work properly?

**Comment by ravisantoshgudimetla** [comment]
Seems, the link is still pointing to old IP address, I will look into this today/tomorrow.

**Comment by krudskoy** [comment]
Hello @ravisantoshgudimetla, are there some changes?

**Comment by ravisantoshgudimetla** [comment]
> Hello @ravisantoshgudimetla, are there any changes?

I have updated the jenkins end point to point to latest ip, I will some restrictions once the test cases pass.

**Comment by krudskoy** [comment]
Thanks

File: pkg/descheduler/strategies/node_affinity.go [hunk 1]
```diff
@@ -26,19 +26,29 @@ import (
 	"k8s.io/api/core/v1"
 )
 
+func Contains(a []string, x string) bool {
```
Comment (ravisantoshgudimetla) [inline pkg/descheduler/strategies/node_affinity.go]: Any specific reason for making this function public?

File: pkg/descheduler/strategies/node_affinity.go [hunk 2]
```diff
@@ -53,21 +63,52 @@ func removePodsViolatingNodeAffinityCount(ds *options.DeschedulerServer, strateg
 					if maxPodsToEvict > 0 && nodepodCount[node]+1 > maxPodsToEvict {
 						break
 					}
-					if pod.Spec.Affinity != nil && pod.Spec.Affinity.NodeAffinity != nil && pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
+					if pod.Spec.Affinity != nil && pod.Spec.Affinity.NodeAffinity != nil {
```
Comment (ravisantoshgudimetla) [inline pkg/descheduler/strategies/node_affinity.go]: You can combine both the blocks under switch cases to function and re-use them.

File: pkg/descheduler/strategies/node_affinity.go (hunk 1 above)
Comment (krudskoy) [reply pkg/descheduler/strategies/node_affinity.go]: Actually no, changed it.

File: pkg/descheduler/strategies/node_affinity.go (hunk 2 above)
Comment (krudskoy) [reply pkg/descheduler/strategies/node_affinity.go]: Got it.

**Comment by yoda** [comment]
> Thank you @krudskoy for the PR. I am curious if you have any use-case for evicting preferred pods too.

@ravisantoshgudimetla Use case would be where you have 2 autoscaler groups of nodes where one is a spot instance type and the other is standard. The spot nodes get terminated on price going above threshold resulting in the preferredDuringScheduling being invalidated resulting in scheduling onto standard nodes. Over time the price on the spots goes back down and the descheduler can cause the rescheduling of them back onto spot instances.

**Comment by seanmalloy** [comment]
@yoda thanks for the feedback. It looks like feature request #211 is currently opened for tracking this. You might want to comment here with any thoughts or feedback. Thanks!

**Pull request closed without merging**